File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/TemplateMethod/concrete_class.rs
// 具体饮料实现
use crate::abstract_class::BeverageMaker;

// 咖啡制作类
pub struct CoffeeMaker;

impl BeverageMaker for CoffeeMaker {
    // 实现冲泡方法
    fn brew(&self) {
        println!("Brewing coffee grounds");
    }
    
    // 实现添加调料方法
    fn add_condiments(&self) {
        println!("Adding sugar and milk");
    }
    
    // 重写钩子方法
    fn customer_wants_condiments(&self) -> bool {
        println!("Customer wants sugar and milk");
        true
    }
}

// 茶制作类
pub struct TeaMaker;

impl BeverageMaker for TeaMaker {
    // 实现冲泡方法
    fn brew(&self) {
        println!("Steeping the tea bag");
    }
    
    // 实现添加调料方法
    fn add_condiments(&self) {
        println!("Adding lemon");
    }
    
    // 重写钩子方法
    fn customer_wants_condiments(&self) -> bool {
        println!("Customer wants lemon");
        true
    }
}

File: 01_GOF23/DP_examples/C/03_Behavioral/TemplateMethod/concrete_class.c
#include "concrete_class.h"

// 咖啡冲泡方法
static void CoffeeMaker_brew(BeverageMaker* self) {
    printf("Brewing coffee grounds\n");
}

// 咖啡添加调料方法
static void CoffeeMaker_addCondiments(BeverageMaker* self) {
    printf("Adding sugar and milk\n");
}

// 咖啡顾客是否需要调料
static int CoffeeMaker_customerWantsCondiments(BeverageMaker* self) {
    printf("Customer wants sugar and milk\n");
    return 1;
}

// 初始化咖啡制作类
void CoffeeMaker_init(CoffeeMaker* self) {
    BeverageMaker_init(&self->base);
    self->base.brew = CoffeeMaker_brew;
    self->base.addCondiments = CoffeeMaker_addCondiments;
    self->base.customerWantsCondiments = CoffeeMaker_customerWantsCondiments;
}

// 茶冲泡方法
static void TeaMaker_brew(BeverageMaker* self) {
    printf("Steeping the tea bag\n");
}

// 茶添加调料方法
static void TeaMaker_addCondiments(BeverageMaker* self) {
    printf("Adding lemon\n");
}

// 茶顾客是否需要调料
static int TeaMaker_customerWantsCondiments(BeverageMaker* self) {
    printf("Customer wants lemon\n");
    return 1;
}

// 初始化茶制作类
void TeaMaker_init(TeaMaker* self) {
    BeverageMaker_init(&self->base);
    self->base.brew = TeaMaker_brew;
    self->base.addCondiments = TeaMaker_addCondiments;
    self->base.customerWantsCondiments = TeaMaker_customerWantsCondiments;
}
